src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -396,19 +390,22 @@ private void StartAsynchronousComputation(AsynchronousComputationToStart computa
                         requestToCompleteSynchronously.CompleteFromTaskSynchronously(task);
                     }
                 }
-                catch (Exception e) when(FatalError.ReportUnlessCanceled(e))
+                catch (Exception e) when (FatalError.ReportUnlessCanceled(e))
                 {
                     throw ExceptionUtilities.Unreachable;
                 }
-                }
-                catch (OperationCanceledException oce) when(CrashIfCanceledWithDifferentToken(oce, cancellationToken))
-                {
-                    // As long as it's the right token, this means that our thread was the first thread
-                    // to start an asynchronous computation, but the requestor cancelled as we were starting up
-                    // the computation.
-                    throw ExceptionUtilities.Unreachable;
-                }
-                }
+            }
+            catch (OperationCanceledException oce) when (CrashIfCanceledWithDifferentToken(oce, cancellationToken))
+            {
+                // The underlying computation cancelled with the correct token, but we must ourselves insure that the caller
+                // on our stack gets an OperationCanceledException thrown with the right token
+                requestToCompleteSynchronouslyCancellationToken.ThrowIfCancellationRequested();
+
+                // The cancellation must have been requested, so this is unreachable
```

Comment:
Comment clarified. To summarize: we only would cancel the underlying computation if all requests for the computation themselves cancelled. In that case, we know that the specific request's token that we're using the thread of must also have been cancelled.